File: AI-LAB/BFS.py
from collections import deque

def bfs(adj_matrix, initial_node, goal_node):
    queue = deque([[initial_node]])  # Initialize the queue with the first path starting from initial state
    visited = set()  # Keep track of visited nodes

    while queue:
        path = queue.popleft()  # Get the frontmost path from queue
        last_node = path[-1]  # Get the last node of this path

        if last_node == goal_node:  # Check if the last node is the goal node
            return path  # Return the path if it is the goal node

        if last_node not in visited:  # Check if the node is not visited in the current path
            visited.add(last_node)  # Mark the node as visited

            # Find all nodes connected to the current node in the adjacency matrix
            connected_nodes = [node for node, connected in enumerate(adj_matrix[last_node]) if connected]

            for node in connected_nodes:
                new_path = list(path)  # Create a new path from the earlier path
                new_path.append(node)  # Append the connected node to the new path
                queue.append(new_path)  # Insert the new path to the queue

    return None  # Return None if no path is found

# Example usage
adj_matrix = [
    [0, 1, 1, 0, 0],
    [1, 0, 0, 1, 0],
    [1, 0, 0, 1, 1],
    [0, 1, 1, 0, 1],
    [0, 0, 1, 1, 0]
]

initial_node = 0
goal_node = 4

optimal_path = bfs(adj_matrix, initial_node, goal_node)
if optimal_path:
    print("Optimal path:", optimal_path)
else:
    print("No path found")

File: AI-LAB/BFS.cpp
#include <iostream>
#include <queue>
#include <vector>

using namespace std;

// Function to perform BFS
void bfs(vector<vector<int>>& graph, int initialNode, int goalNode) {
    int numNodes = graph.size();
    vector<bool> visited(numNodes, false);
    queue<vector<int>> paths;

    // Initialize the queue with the first path starting from initial state
    paths.push({initialNode});

    while (!paths.empty()) {
        // Get the frontmost path from the queue
        vector<int> path = paths.front();
        paths.pop();

        // Get the last node of the current path
        int lastNode = path.back();

        // Check if the last node is the goal node
        if (lastNode == goalNode) {
            // Print the path
            for (int node : path) {
                cout << node << " ";
            }
            cout << endl;
            return;
        }

        // Check all vertices connected to the current node
        for (int i = 0; i < numNodes; i++) {
            if (graph[lastNode][i] == 1 && !visited[i]) {
                // Create a new path from the earlier path and append the vertex
                vector<int> newPath = path;
                newPath.push_back(i);

                // Insert the new path to the queue
                paths.push(newPath);

                // Mark the vertex as visited
                visited[i] = true;
            }
        }
    }
}

int main() {
    // Example adjacency matrix
    vector<vector<int>> graph = {
        {0, 1, 1, 0, 0},
        {1, 0, 0, 1, 0},
        {1, 0, 0, 1, 1},
        {0, 1, 1, 0, 1},
        {0, 0, 1, 1, 0}
    };

    int initialNode = 0;
    int goalNode = 4;

    bfs(graph, initialNode, goalNode);

    return 0;
}